src/NUnitFramework/framework/Internal/Execution/WorkItemBuilder.cs

```diff
@@ -56,42 +56,52 @@ static public WorkItem CreateWorkItem(ITest test, ITestFilter filter, bool recur
         /// <param name="filter">The filter to be used in selecting any child Tests.</param>
         /// <param name="debugger">An <see cref="IDebugger" /> instance.</param>
         /// <param name="recursive">True if child work items should be created and added.</param>
+        /// <param name="root">True iа work item needs to be created unconditionally, if false  null would be returned for tests that don't match the filter</param>
         /// <returns></returns>
-        internal static WorkItem CreateWorkItem(ITest test, ITestFilter filter, IDebugger debugger, bool recursive = false)
+        internal static WorkItem CreateWorkItem(ITest test, ITestFilter filter, IDebugger debugger, bool recursive = false, bool root = true)
         {
+            // Run filter on leaf nodes only
+            // use the presense of leaf nodes as an indicator that parent need to be created
+            // Always create a workitem for the root node
             TestSuite suite = test as TestSuite;
             if (suite == null)
-                return new SimpleWorkItem((TestMethod)test, filter, debugger);
+            {
+                if (root || filter.Pass(test))
+                {
+                    return new SimpleWorkItem((TestMethod)test, filter, debugger);
+                }
+                return null;
+            }
 
-            var work = new CompositeWorkItem(suite, filter);
+            CompositeWorkItem work = root ? new CompositeWorkItem(suite, filter): null;
 
             if (recursive)
             {
                 int countOrderedItems = 0;
 
                 foreach (var childTest in suite.Tests)
                 {
-                    if (filter.Pass(childTest))
+                    var childItem = CreateWorkItem(childTest, filter, debugger, recursive: true, root: false);
```

Comment:
Perhaps we should pass in `recursive` instead of `true`, to be more robust if we at some point make a small change to the expression in the if-statement.